src/plugin/file/zip/ui/zipimport.js

```diff
@@ -0,0 +1,347 @@
+   * @private
+   */
+  this.importers_;
+
+  /**
+   * Group of objects related to each step in the chain of importers
+   *
+   * @type {plugin.file.zip.ImporterPair|null}
+   * @private
+   */
+  this.curImporter_ = null;
+
+  /**
+   * Group of objects related to each step in the chain of importers
+   *
+   * @type {function()|null}
+   * @private
+   */
+  this.curListener_ = null;
+
+  /**
+   * Group of objects related to each step in the chain of importers
+   *
+   * @type {angular.$q.Promise|null}
+   * @private
+   */
+  this.curTimeout_ = null;
+
+  /**
+   * @type {plugin.file.zip.ZIPParserConfig}
+   * @private
+   */
+  this.config_ = /** @type {plugin.file.zip.ZIPParserConfig} */ ($scope['config']);
+
+  /**
+   * @type {number}
+   * @private
+   */
+  this.wait_ = 0;
+
+  /**
+   * @type {Array.<osx.import.FileWrapper>|null}
+   */
+  this['files'] = this.config_['files'];
+
+  /**
+   * @type {boolean}
+   */
+  this['loading'] = false;
+
+  /**
+   * @type {boolean}
+   */
+  this['valid'] = false;
+
+  this.init_();
+};
+goog.inherits(plugin.file.zip.ui.ZIPImportCtrl, os.ui.file.FileImportCtrl);
+
+
+/**
+ * @export
+ */
+plugin.file.zip.ui.ZIPImportCtrl.prototype.finish = function() {
+  var entries = this['files'];
+
+  if (!entries) {
+    var msg = 'No files selected to import from ZIP';
+    os.alert.AlertManager.getInstance().sendAlert(msg, os.alert.AlertEventSeverity.WARNING);
+    return;
+  }
+
+  this.importers_ = [];
+  var unsupported = {};
+
+  for (var i = 0; i < entries.length; i++) {
+    var entry = entries[i];
+    if (entry.enabled === true) {
+      var file = entry.file;
+      var type = file.getType();
+      var ui = this.im_.getImportUI(type);
+      if (ui) {
+        var importer = /** @type plugin.file.zip.ImporterPair */ ({file, ui});
+        this.importers_.push(importer);
+      } else unsupported[type] = true; // store the filetype to report to user later
+    }
+  }
+
+  var keys = Object.keys(unsupported);
+
+  if (keys && keys.length > 0) {
+    var err = 'Unsupported filetype(s).<br />' + keys.join(', ');
+    os.alert.AlertManager.getInstance().sendAlert(err, os.alert.AlertEventSeverity.ERROR);
+  }
+
+  // either close after the chain completes, or close since there is no chain
+  if (this.importers_ && this.importers_.length > 0) this.chain();
+  else this.close(); // for now, close this and rely on the chain; possible future improvement: use slickgrid and show import status
+};
+
+
+/**
+ * Kick off the import of an unzipped file
+ */
+plugin.file.zip.ui.ZIPImportCtrl.prototype.chain = function() {
+  if (typeof this.importers_ == 'undefined' || this.importers_ == null || this.importers_.length == 0) {
+    if (this.curListener_) this.curListener_(); // kill the listener that was created by the previous chain()
+    this.close();
+    return;
+  }
+
+  var onSuccess = function() {
+    // TODO if there are other paths through importers in the future, then update this code
```

Comment:
This is a warning to future developers -- This part of the code is a bit meh, as it only works "nicely" with os.ui.Windows that follow our Importer scope data structure.  Other types of popups or modal frameworks don't throw the same events, so the chain() will just kick off the next importer after 1.35 seconds so least all the file(s) they want will get imported, but it may be annoying to users to have more than one popup going at a time.  

The next TODO talking about making a breaking change to the Importers to enforce that they all throw events would fix this weakness.  That will likely be turned into a follow-on ticket.